### DISCUSSION TITLE
[Feature] TokenView

### DISCUSSION THREAD

**Discussion by niels9001** [CommunityToolkit/Labs-Windows#323] [Proposals]
# TokenView

## Problem Statement

Often, you'd like to present e.g. selectable options to users. Some examples: UI to filter a list, present contacts or other options. It can be cumbersome to tweak a List/GridView to achieve this - an out of the box control that presents options consistently using the W11 design language would be a timesaver!

Android has the ChipControl for this: https://m3.material.io/components/chips/overview 

## Inspiration

In W11 Search, this patterns is used to allow the user to filter search results:

![Filter](https://user-images.githubusercontent.com/9866362/204153276-17e88b27-e665-4061-972e-5ea7fabfe956.gif)

In the W11 Store, there's a similiar control (although it has a different look and feel):
![image](https://user-images.githubusercontent.com/9866362/204153089-3dd1efd8-f324-42ee-b58d-b8e1f138ded7.png)

## Proposed Solution
A lightweight control that allows for setting different tokens deriving off ListView. Additionally, multi-select can be enabled. It should adhere to the Windows Visual Design specs so it feels native to WinUI. There should be scroll buttons (instead of a scrollbar) that animates to the first or last item, similar to W11 Search.

```xml
 <TokenView>
     <TokenItem Content="All" />
     <TokenItem Content="Work" />
     <TokenItem Content="Apps" />
     <TokenItem Content="Documents" />
     <TokenItem Content="Web" />
     <TokenItem Content="Settings" />
     <TokenItem Content="People" />
     <TokenItem Content="Folders" />
     <TokenItem Content="Photos" />
</TokenView>
```

result:
![image](https://user-images.githubusercontent.com/9866362/204153222-f9fcf654-6b22-42d7-988c-97bd378103e4.png)

Properties for TokenView:

| Property name  | Type | Comments |
| ------------- | ------------- | ------------- |
| Orientation | TokenViewOrientation | Can be Horizontal (Default) or Wrapped. In Horizontal mode, scrollbuttons will appears. In Wrapped mode, TokenItems will be wrapped horizontally + vertically.
| CanRemoveTokens | bool | This makes all child TokenItems removeable.
| SelectionMode | ListViewSelectionMode | Allows for single / multi-selection

Properties for TokenViewItem:

| Property name  | Type | Comments |
| ------------- | ------------- | ------------- |
| Icon | IconElement | Icon showing left of the Content
| Content | object | Content of the TokenItem
| IsRemoveable| bool | Determines if a TokenItem can be removed (using Del/Backspace or the remove button.
| SelectionMode | ListViewSelectionMode | Allows for single / multi-selection

## References
- https://www.figma.com/file/XjuZJySxR0ydkkYIrqk7ps/Windows-UI-3-(Community)?node-id=55073%3A246118&t=z6U37IDcuHdal6xA-4

**Comment by michael-hawker** [comment]
@niels9001 I think the general form of this would be the _segmented control_, eh? See WinUI proposal here with some prior discussion: https://github.com/microsoft/microsoft-ui-xaml/issues/2310 - Also came up when looking at `RadioButtons` for similar experience, as I discussed here: https://github.com/microsoft/microsoft-ui-xaml/issues/3759

**Comment by michael-hawker** [comment]
I think the common term for this control is a "Chips" control, eh? See original thread on Community Toolkit here: https://github.com/CommunityToolkit/WindowsCommunityToolkit/issues/1738 - which led to the TokenizingTextBox (which would be the editable version I guess?)

See Material docs here for the control (covers the input scenarios as well): https://www.material.io/components/chips/android
